sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_oneapi_matrix.asciidoc

```diff
@@ -427,6 +433,37 @@ joint_matrix_apply(sg, C, [=](T &x) {
 });
 ```
 
+===== Input and output matrix are different
+In this case, `joint_matrix_apply` takes two `joint_matrix` arguments:
+`srcjm` and `destjm` that have the same `use`, type, number of rows,
+number of columns, and `layout`. `srcjm` is a read-only argument but
+`destjm` is a read and write argument. The callable object must be
+invocable with two parameters of type `T&`. Commonly, applications
+pass a lambda expression.
+
+```c++
+namespace sycl::ext::oneapi::experimental::matrix {
+
+template<typename Group, typename T, use Use, size_t Rows, size_t Cols,
+  layout Layout, typename F>
+void joint_matrix_apply(Group g,
+          joint_matrix<Group, T, Use, Rows, Cols, Layout>& srcjm,
+          joint_matrix<Group, T, Use, Rows, Cols, Layout>& destjm,
+	  F&& func);
+
+} // namespace sycl::ext::oneapi::experimental::matrix
+```
+
+In the following example, every element of the matrix `C` is
+multiplied by `alpha`. The result is returned into a different matrix
+`D`.
+
+```c++
+joint_matrix_apply(sg, C, D, [=](const T &x, T &y) {
+    y = x * alpha;
+});
+```
```

Comment:
I don't see a problem with this API, but I wonder if it should be more general.  For example:

* Is there a reason that one matrix must be the "source" and the other the "destination"?  Is it hard to support the case where the lambda writes to elements in both matrices?

* I think it makes sense that `Rows` and `Cols` must be the same for both matrices.  What about `Use` and `Layout`, though?  Is there a reason these can't be different for the two matrices?  What about the type `T`?  Could that be different?

* Is there a reason to limit this to just two matrices?  You could imagine this API taking a parameter pack, allowing an arbitrary number of matrices.

Reply:
@gmlueck,

- I relaxed the definition to not require read or write assumptions. Both matrices can be either read or written into.

- I think it is safer to require the same `use` and `layout`. To change the `use`, `joint_matrix_copy` can be used. 
`T` should be the same. This is because there are different casts that can be used depending on `T`. It is better to have the user specify the right cast in the lambda.

- Using a parameter pack will require the lambda to also use a parameter pack. Also, how do we pack the wi_data elements and pass them to the lambda (see line 22 in https://godbolt.org/z/E1sT5oY3P).
I don' think passing a variadic number will be useful as the user can implement any number of arguments with the two arguments version. As an optimization, we can also add a three arguments version to avoid using intermediate matrices in the case of ternary operation like C=A*B.

What do you think?

Reply:
> * I think it is safer to require the same `use` and `layout`. To change the `use` or `layout`, `joint_matrix_copy` can be used https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_oneapi_matrix.asciidoc#copy.
>   `T` should be the same. This is because there are different casts that can be used depending on `T`. It is better to have the user specify the right cast in the lambda.

I don't understand why it is "safer" to require `use`, `layout`, and `T` to be the same for the two matrices.  In what sense is it safer?  What use cases do you have in mind for the form that takes two matrices?  For these use cases, will it always be the case that `use`, `layout`, and `T` are the same?

Maybe there is an implementation problem.  The current implementation probably assumes that the distribution of matrix elements across work-items will be the same for the two matrices.  For example, I suspect your implementation assumes that `wi_data[j]` for work-item `I` has the same _Row_ / _Col_ coordinate in both the _A_ and _B_ matrices.  Does this assumption become invalidated if `use`, `layout`, or `T` are different for _A_ and _B_?

Reply:
Correct, if use, layout and T are different, there is no guarantee work items will have same number of data. 
Yes, the CUDA use cases suggest use, layout and T are the same.

Reply:
@gmlueck, as a correction, if use and layout are different, there is not guarantee of ownership.
So changing type should be fine.
I will relax the type restriction to enable the use cases where conversion with extra arguments is needed.